server/src/main/java/org/elasticsearch/cluster/project/ProjectStateRegistry.java

```diff
@@ -116,22 +171,47 @@ public static Builder builder() {
 
     public static class Builder {
         private final ImmutableOpenMap.Builder<ProjectId, Settings> projectsSettings;
+        private final Set<ProjectId> projectsMarkedForDeletion;
+        private final long projectsMarkedForDeletionGeneration;
+        private boolean newProjectMarkedForDeletion = false;
 
         private Builder() {
             this.projectsSettings = ImmutableOpenMap.builder();
+            projectsMarkedForDeletion = new HashSet<>();
+            projectsMarkedForDeletionGeneration = 0;
         }
 
         private Builder(ProjectStateRegistry original) {
             this.projectsSettings = ImmutableOpenMap.builder(original.projectsSettings);
+            this.projectsMarkedForDeletion = new HashSet<>(original.projectsMarkedForDeletion);
+            this.projectsMarkedForDeletionGeneration = original.projectsMarkedForDeletionGeneration;
         }
 
         public Builder putProjectSettings(ProjectId projectId, Settings settings) {
             projectsSettings.put(projectId, settings);
             return this;
         }
 
+        public Builder markProjectForDeletion(ProjectId projectId) {
+            if (projectsMarkedForDeletion.add(projectId)) {
+                newProjectMarkedForDeletion = true;
+            }
+            return this;
+        }
+
         public ProjectStateRegistry build() {
-            return new ProjectStateRegistry(projectsSettings.build());
+            final var unknownButUnderDeletion = Sets.difference(projectsMarkedForDeletion, projectsSettings.keys());
+            if (unknownButUnderDeletion.isEmpty() == false) {
+                assert unknownButUnderDeletion.stream().noneMatch(projectsSettings::containsKey);
```

Comment:
What I had in mind is something like `assert false : "Cannot mark projects ..."`. On a second thought, an invalid file-based configuration can actually trigger it so that it is not only a programming error. Therefore, I think we don't really want the assertion that I had mind. We also don't need this assertion here since it just asserts the correctness of the above `Sets.difference` call. Sorry for the back and forth.